Switched the BlockEvent stream to Tokio broadcast

neonknight64 opened the pull request:
## Description
- The BlockEvent stream currently uses the tari_broadcast_channel, this channel is bounded and does not clear when the receivers have read the events. As newly added and reorged blocks are distributed between services using this event stream, the channel can be memory intensive. 
- These changes replaced the tari_broadcast_channel used for the BlockEvent stream with a tokio broadcast channel. The Tokio broadcast channel has the benefit of removing the events from the channel as soon as all the receivers have read the events. This should minimise the amount of memory required.

## Motivation and Context
The tari_broadcast_channel is memory intensive for sharing BlockEvents as it keeps the full set of events up to the specified bound limit even when the receivers have already processed the events. Using a tokio broadcast channel will result in less memory usage as all fully received events will be discarded without reaching the bound limit.

## How Has This Been Tested?
Covered by existing tests.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
* [x] Bug fix (non-breaking change which fixes an issue)
* [ ] New feature (non-breaking change which adds functionality)
* [ ] Breaking change (fix or feature that would cause existing functionality to change)
* [x] Feature refactor (No new feature or functional changes, but performance or technical debt improvements)
* [ ] New Tests
* [ ] Documentation

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
* [x] I'm merging against the `development` branch.
* [x] I ran `cargo-fmt --all` before pushing.
* [x] I have squashed my commits into a single commit.
* [ ] My change requires a change to the documentation.
* [ ] I have updated the documentation accordingly.
* [ ] I have added tests to cover my changes.

philipr-za reviewed (dismissed):
LGTM

sdbondi commented on base_layer/core/src/base_node/comms_interface/local_interface.rs:
```diff
@@ -33,39 +33,42 @@ use crate::{
+use tokio::sync::broadcast;
 use tower_service::Service;
 
+pub type BlockEventSender = broadcast::Sender<BlockEvent>;
```
tokio broadcast assumes that the item broadcast is cheap to clone. BlockEvent has a large memory footprint and is expensive to clone, so I think we should wrap it in an Arc.

neonknight64 replied:
Good suggestion, I wrapped the BlockEvent in an Arc.

sdbondi reviewed (changes requested):
Looks good - but BlockEvent is too expensive to clone, so should be wrapped in an Arc when broadcasting.

Commit:
- The tari_broadcast_channel is memory intensive for sharing BlockEvents as it keeps the full set of events up to the specified bound limit even when the receivers have already processed the events. Using a tokio broadcast channel will result in less memory usage as all fully received events will be discarded without reaching the bound limit.
- The BlockEvent stream currently uses the tari_broadcast_channel, this channel is bounded and does not clear when the receivers have read the events. As newly added and reorged blocks are distributed between services using this event stream, the channel can be memory intensive.
- These changes replaced the tari_broadcast_channel used for the BlockEvent stream with a tokio broadcast channel. The Tokio broadcast channel has the benefit of removing the events from the channel as soon as all the receivers have read the events. This should minimise the amount of memory required.
